### ISSUE TITLE
Single-Tap on non-empty TextField, VoiceOver says "empty line"

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#52487]
See b/151307855

Observe that when using `TextField` or `TextFormField` in Flutter on iOS platform, if users Single-Tap a non-empty, `TextField` (the text field is not in an active/editing state) under VoiceOver mode, it will read out 'empty line'.

Steps to reproduce:

Turn on VoiceOver
Navigate to a screen that has a TextField or TextFormField

case 1: Navigate to the TextField using __swipe__ gestures, it voice out 
- `Text field, is editing, <content>`

case 2: __Single-Tap__ on TextField directly, it voice out 
- `Text field, empty line`

**Comment by goderbauer** [comment]
Would be good to get a print-out of the semantics tree as well to see what information we are providing to VoiceOver.

**Comment by Piinks** [comment]
Clarified expected behavior on internal bug, expected to always reflect case 1.

So far I have not been able to reproduce this. Here are my findings:

---

### Flutter Doctor
```
[✓] Flutter (Channel master, v1.15.22-pre.11, on Mac OS X 10.15.3 19D76, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
[✓] Android Studio (version 3.6)
[!] IntelliJ IDEA Ultimate Edition (version 2019.2.4)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.42.1)
[✓] Connected device (1 available)
```

---

### Code Sample

```dart
// Flutter code sample for Form

// This example shows a [Form] with one [TextFormField] to enter an email
// address and a [RaisedButton] to submit the form. A [GlobalKey] is used here
// to identify the [Form] and validate input.
//
// ![](https://flutter.github.io/assets-for-api-docs/assets/widgets/form.png)

import 'package:flutter/material.dart';

void main() => runApp(MyApp());

/// This Widget is the main application widget.
class MyApp extends StatelessWidget {
  static const String _title = 'Flutter Code Sample';

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: _title,
      home: Scaffold(
        appBar: AppBar(title: const Text(_title)),
        body: MyStatefulWidget(),
      ),
    );
  }
}

class MyStatefulWidget extends StatefulWidget {
  MyStatefulWidget({Key key}) : super(key: key);

  @override
  _MyStatefulWidgetState createState() => _MyStatefulWidgetState();
}

class _MyStatefulWidgetState extends State<MyStatefulWidget> {
  final _formKey = GlobalKey<FormState>();

  @override
  Widget build(BuildContext context) {
    return Form(
      key: _formKey,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: <Widget>[
          TextFormField(
            validator: (value) {
              if (value.isEmpty) {
                return 'Please enter some text';
              }
              return null;
            },
          ),
          Padding(
            padding: const EdgeInsets.symmetric(vertical: 16.0),
            child: RaisedButton(
              onPressed: () {
                // Validate will return true if the form is valid, or false if
                // the form is invalid.
                if (_formKey.currentState.validate()) {
                  // Process data.
                }
              },
              child: Text('Submit'),
            ),
          ),
        ],
      ),
    );
  }
}

```

---

### iOS Testing
iPhone 11 Pro, iOS 13.3.1

#### Native Behavior (Safari)

_Swiping to text field_
- 'Text field, double tap to edit'
- :double taps, keyboard presents:
- 'Insertion point at end'
- :swipes from there and back, keyboard still open:
- 'Text field, is editing, insertion point at start

_Tapping on text field_
- 'Text field, double tap to edit'
- :double taps, keyboard presents:
- 'Insertion point at end'
- :taps on text field, keyboard still open:
- 'Text field, is editing, insertion point at start

#### Flutter Behavior

_Swiping to text field_
- 'Text field, double tap to edit'
- :double taps, keyboard presents:
- 'Text field, is editing, insertion point at start'
- :swipes away and back, keyboard still open:
- 'Text field, is editing, character mode, insertion point at start'

_Tapping on text field_
- 'Text field, double tap to edit'
- :double taps, keyboard presents:
- 'Text field, is editing, insertion point at start'
- :taps on text field, keyboard still open:
- 'Text field, is editing, character mode, insertion point at start'

#### Semantics Tree
```
flutter: SemanticsNode#0
flutter:  │ Rect.fromLTRB(0.0, 0.0, 1125.0, 2436.0)
flutter:  │
flutter:  └─SemanticsNode#1
flutter:    │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0) scaled by 3.0x
flutter:    │ textDirection: ltr
flutter:    │
flutter:    └─SemanticsNode#2
flutter:      │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:      │ flags: scopesRoute
flutter:      │
flutter:      ├─SemanticsNode#5
flutter:      │ │ Rect.fromLTRB(0.0, 0.0, 375.0, 100.0)
flutter:      │ │ thickness: 4.0
flutter:      │ │
flutter:      │ └─SemanticsNode#6
flutter:      │     Rect.fromLTRB(94.5, 61.0, 280.5, 83.0)
flutter:      │     flags: isHeader
flutter:      │     label: "Flutter Code Sample"
flutter:      │     textDirection: ltr
flutter:      │     elevation: 4.0
flutter:      │
flutter:      ├─SemanticsNode#3
flutter:      │   Rect.fromLTRB(0.0, 100.0, 375.0, 148.0)
flutter:      │   actions: tap
flutter:      │   flags: isTextField
flutter:      │   textDirection: ltr
flutter:      │   currentValueLength: 0
flutter:      │
flutter:      └─SemanticsNode#4
flutter:          Rect.fromLTRB(0.0, 164.0, 88.0, 212.0)
flutter:          actions: tap
flutter:          flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          label: "Submit"
flutter:          textDirection: ltr
flutter:          thickness: 2.0
flutter:
```
I have asked for a reproduction of the reported issue on the internal bug.

**Comment by Piinks** [comment]
I have been able to reproduce, this happens when text is entered into the text field, and when the text field is not active for editing, tapping on it reads out 'empty line'.

Here is the semantics tree when the above sample contains text in the text field:

```
flutter: SemanticsNode#0
flutter:  │ Rect.fromLTRB(0.0, 0.0, 1125.0, 2436.0)
flutter:  │
flutter:  └─SemanticsNode#7
flutter:    │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0) scaled by 3.0x
flutter:    │ textDirection: ltr
flutter:    │
flutter:    └─SemanticsNode#8
flutter:      │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:      │ flags: scopesRoute
flutter:      │
flutter:      ├─SemanticsNode#11
flutter:      │ │ Rect.fromLTRB(0.0, 0.0, 375.0, 100.0)
flutter:      │ │ thickness: 4.0
flutter:      │ │
flutter:      │ └─SemanticsNode#12
flutter:      │     Rect.fromLTRB(94.5, 61.0, 280.5, 83.0)
flutter:      │     flags: isHeader
flutter:      │     label: "Flutter Code Sample"
flutter:      │     textDirection: ltr
flutter:      │     elevation: 4.0
flutter:      │
flutter:      ├─SemanticsNode#9
flutter:      │   Rect.fromLTRB(0.0, 100.0, 375.0, 148.0)
flutter:      │   actions: tap
flutter:      │   flags: isTextField
flutter:      │   value: "entered text"
flutter:      │   textDirection: ltr
flutter:      │   currentValueLength: 12
flutter:      │
flutter:      └─SemanticsNode#10
flutter:          Rect.fromLTRB(0.0, 164.0, 88.0, 212.0)
flutter:          actions: tap
flutter:          flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          label: "Submit"
flutter:          textDirection: ltr
flutter:          thickness: 2.0
flutter:
```

**Comment by Piinks** [comment]
Since I figured out how to reproduce on iOS, I double checked Android, this does not reproduce on Android.
I think this is a bug in the ios accessibility bridge in the engine.

**Comment by Piinks** [comment]
cc/ @xster for the platforms teams. :)

**Comment by xster** [comment]
Thanks for doing the analysis @Piinks!

**Comment by goderbauer** [comment]
I wonder if something goes wrong in the a11y bridge where we swap out the SemanticsNode for an actual UiTextField to steal its semantics.

**Comment by goderbauer** [comment]
@gaaclarke I was told that you may be looking at this? Assigned it to you. Please unassign if this was wrong.

**Comment by gaaclarke** [comment]
- The problem is most likely in accessibility_text_entry.mm:
https://github.com/flutter/engine/blob/857859f8018f9b9b99b0adca978a8198350176f7/shell/platform/darwin/ios/framework/Source/accessibility_text_entry.mm#L254
- Hardcoding `accessibilityLabel` to return "label" and `accessibilityValue` to return "value", single tapping the textfield says "Label, empty line".
- TextInputSemanticsObject is implemented differently than I implemented [FlutterSwitchSemanticsObject ](https://github.com/flutter/engine/blob/857859f8018f9b9b99b0adca978a8198350176f7/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L79). It forwards the methods explicitly instead of hijacking method resolution which means it might be missing weird reflection / private methods VoiceOver is looking for.
- just to to be clear, the expected behavior should be VoiceOver saying "<contents>, TextField, Double-tap to edit". (verified with a minimal Xcode project)

**Comment by gaaclarke** [comment]
- If you hardcode `-[TextInputSemanticsObject textInputSurrogate]` to always return `[self bridge] -> textInputView();`, the bug still manifests.  I was wondering if this was some difference in state between the bridges input view and the `_inactive_text_input`.

**Comment by gaaclarke** [comment]
- Hardcoding `-[TextInputSemanticsObject accessibilityTraits]` to always return zero causes VoiceOver to not say "Empty Line"
- Printing out the results to `-[TextInputSemanticsObject accessibilityTraits]`, we see that in the case where it works (swiping) vs the pathological case (tapping) the results are exactly the same: `[VERBOSE-2:accessibility_text_entry.mm(289)] traits:0x802000040020`

**Comment by gaaclarke** [comment]
- 0x802000040020 maps to only one documented trait: UIAccessibilityTraitKeyboardKey
- forcing traits to return `0x802000040020` causes it to say "empty line"
- returning `0x002000040020` removes the bug, the problem is that `0x800000000000` appears to be undocumented...
- Here is a diff that fixes the issues, still looking for more solid solution:
```
--- a/shell/platform/darwin/ios/framework/Source/accessibility_text_entry.mm
+++ b/shell/platform/darwin/ios/framework/Source/accessibility_text_entry.mm
@@ -284,8 +284,9 @@
   // Adding UIAccessibilityTraitKeyboardKey to the trait list so that iOS treats it like
   // a keyboard entry control, thus adding support for text editing features, such as
   // pinch to select text, and up/down fling to move cursor.
-  return [super accessibilityTraits] | [self textInputSurrogate].accessibilityTraits |
+  UIAccessibilityTraits results = [super accessibilityTraits] | [self textInputSurrogate].accessibilityTraits |
          UIAccessibilityTraitKeyboardKey;
+  return results & (~0x800000000000);
 }
```

**Comment by gaaclarke** [comment]
I already asking on the b/ link, but can someone here test on iOS 12?  The reporter only reported it for iOS 13 and I don't have anything but an iOS 13 device available.

**Comment by gaaclarke** [comment]
@goderbauer what's our feeling on squashing the `0x800000000000` flag considering empirically it works but we don't know exactly what it represents?  Have we had to do this in the past?

I think the only hope of an alternative is to make TextInputSemanticsObject a UITextField which would be a pretty big change because of the way we set things up with the bridge to share a text input field.
